### PR TITLE
Forum channels support

### PR DESCRIPTION
**Pull request by BobDotCom**
## Summary

Implements forum channels

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

File: discord/abc.py [hunk 1]
```diff
@@ -81,7 +81,7 @@
         GroupChannel,
         PartialMessageable,
         TextChannel,
-        VoiceChannel,
+        VoiceChannel
```
Comment (plun1331) [inline discord/abc.py:84]: Was this supposed to change? (missing trailing comma)

File: discord/channel.py [hunk 2]
```diff
@@ -75,6 +77,7 @@
     "CategoryChannel",
     "GroupChannel",
     "PartialMessageable",
+    "ForumChannel"
```
Comment (plun1331) [inline discord/channel.py:80]: Missing trailing comma

File: discord/channel.py [hunk 3]
```diff
@@ -169,25 +173,22 @@ class TextChannel(discord.abc.Messageable, discord.abc.GuildChannel, Hashable):
...
-            ("nsfw", self.nsfw),
-            ("news", self.is_news()),
-            ("category_id", self.category_id),
-        ]
+        attrs = [(val, getattr(self, val)) for val in self._repr_attrs]
         joined = " ".join("%s=%r" % t for t in attrs)
         return f"<{self.__class__.__name__} {joined}>"
 
-    def _update(self, guild: Guild, data: TextChannelPayload) -> None:
+    def _update(self, guild: Guild, data: TextChannelPayload | ForumChannelPayload) -> None:
```
Comment (plun1331) [inline discord/channel.py:191]: iirc `|` for union types is only available in 3.10

```suggestion
    def _update(self, guild: Guild, data: Union[TextChannelPayload, ForumChannelPayload]) -> None:
```

File: discord/channel.py [hunk 4]
```diff
@@ -724,55 +787,177 @@ async def create_thread(
...
+    @property
+    def guidelines(self) -> Optional[str]:
+        """Optional[:class:`str`]: The channel's guidelines. An alias of :attr:`topic`.
+        """
+        return self.topic
+
+    async def create_post(
+            self,
+            name: str,  # Could be renamed to title?
+            message_content=None,
```
Comment (plun1331) [inline discord/channel.py:807]: Maybe rename to just `content`?

File: discord/abc.py (hunk 1 above)
Comment (Lulalaby) [inline discord/abc.py:84]:
```suggestion
        VoiceChannel,
```

File: discord/channel.py (hunk 2 above)
Comment (Lulalaby) [inline discord/channel.py:80]:
```suggestion
    "ForumChannel",
```

File: discord/channel.py (hunk 4 above)
Comment (Lulalaby) [inline discord/channel.py:807]:
```suggestion
            content=None,
```

File: discord/channel.py [hunk 5]
```diff
@@ -724,55 +787,177 @@ async def create_thread(
...
-        return ArchivedThreadIterator(
-            self.id,
-            self.guild,
-            limit=limit,
-            joined=joined,
-            private=private,
-            before=before,
-        )
+        state = self._state
+        content = str(message_content) if message_content is not None else None
```
Comment (Lulalaby) [inline discord/channel.py:862]:
```suggestion
        message_content = str(content) if content is not None else None
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

File: discord/channel.py [hunk 6]
```diff
@@ -724,55 +787,177 @@ async def create_thread(
...
+        if file is not None:
+            if not isinstance(file, File):
+                raise InvalidArgument("file parameter must be File")
+
+            try:
+                data = await state.http.send_files(
+                    self.id,
+                    files=[file],
+                    allowed_mentions=allowed_mentions,
+                    content=content,
```
Comment (Lulalaby) [inline discord/channel.py:905]:
```suggestion
                    content=message_content,
```

File: discord/channel.py [hunk 7]
```diff
@@ -724,55 +787,177 @@ async def create_thread(
...
+            if len(files) > 10:
+                raise InvalidArgument("files parameter must be a list of up to 10 elements")
+            elif not all(isinstance(file, File) for file in files):
+                raise InvalidArgument("files parameter must be a list of File")
+
+            try:
+                data = await state.http.send_files(
+                    self.id,
+                    files=files,
+                    content=content,
```
Comment (Lulalaby) [inline discord/channel.py:926]:
```suggestion
                    content=message_content,
```

File: discord/channel.py [hunk 8]
```diff
@@ -724,55 +787,177 @@ async def create_thread(
...
+                    stickers=stickers,
+                    components=components,
+                )
+            finally:
+                for f in files:
+                    f.close()
+        else:
+            data = await state.http.start_forum_thread(
+                self.id,
+                content,
```
Comment (Lulalaby) [inline discord/channel.py:941]:
```suggestion
                content=message_content,
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

**Comment by Lulalaby** [comment]
@plun1331 you mean like that? (Review)

**Review by Lulalaby — APPROVED** [review]

**Comment by plun1331** [comment]
> @plun1331 you mean like that? (Review)

looks fine yeah

**Comment by Lulalaby** [comment]
@Pycord-Development/contributors 
@Pycord-Development/maintainers 

Someone needs to finish this for bob

**Comment by plun1331** [comment]
> @Pycord-Development/contributors @Pycord-Development/maintainers
> 
> Someone needs to finish this for bob

List of what needs to be finished would be nice

**Comment by Lulalaby** [comment]
Check dms

**Comment by ConchDev** [comment]
Github has DMs?

**Comment by plun1331** [comment]
> Github has DMs?

no, discord does though

File: discord/flags.py [hunk 9]
```diff
@@ -1198,3 +1198,39 @@ def gateway_message_content_limited(self):
         and has hit the guild limit.
         """
         return 1 << 19
+
+
+@fill_with_flags()
+class ChannelFlags(BaseFlags):
+    r"""Wraps up the Discord Channel flags.
```
Comment (Lulalaby) [inline discord/flags.py:1205]: Why the r"""

Comment (plun1331) [reply discord/flags.py:1205]: I just copy/pasted from another flag class, they're all like that (except for MemberCacheFlags)

Comment (krittick) [reply discord/flags.py:1205]: It treats the docstring as a raw string (i.e. escape characters are ignored and treated as text). It's fine to leave as-is.

File: discord/channel.py [hunk 10]
```diff
@@ -724,55 +794,198 @@ async def create_thread(
...
+                )
+            finally:
+                for f in files:
+                    f.close()
+        else:
+            data = await state.http.start_forum_thread(
+                self.id,
+                content=message_content,
+                name=name,
+                tts=tts,
```
Comment (krittick) [inline discord/channel.py:970]:
```suggestion
```

**Review by krittick — APPROVED** [review]

**Review by Dorukyum — APPROVED** [review]

**Pull request merged**